### PR TITLE
PRI-32: Screening Questions Redirects

### PR DESCRIPTION
**Pull request by LucaDelBuonoHMCTS**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/PRI-32
(previously https://tools.hmcts.net/jira/browse/PRO-3082)

### Change description ###
Forces the user to answer all the screening questions by setting a cookie and keeping track of what questions have been asked

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[X] No
```

### CODE REVIEW

File: test/unit/testDeceasedDob.js [hunk 1]
```diff
@@ -14,27 +14,49 @@ describe('DeceasedDob', () => {
...
 
-    describe('handlePost()', () => {
-        let ctx;
-        let errors;
-        let formdata;
-        let session;
-        let hostname;
-        let featureToggles;
+    describe('getContextData()', () => {
+        it('should return the ctx with the deceased address and the screening_question feature toggle on', (done) => {
```
Comment (shepda) [inline test/unit/testDeceasedDob.js]: Please update the description for this test as it is for deceased dob but it is referring to deceased address.

File: test/unit/testDeceasedDob.js [hunk 2]
```diff
@@ -14,27 +14,49 @@ describe('DeceasedDob', () => {
...
+            });
+            done();
+        });
 
-        it('should return the ctx with the deceased dob and the screening_question feature toggle', (done) => {
-            ctx = {
-                dob_day: '02',
-                dob_month: '03',
-                dob_year: '1952'
+        it('should return the ctx with the deceased address and the screening_question feature toggle off', (done) => {
```
Comment (shepda) [inline test/unit/testDeceasedDob.js]: Please update the description for this test as it is for deceased dob but it is referring to deceased address.

File: test/unit/testDeceasedDod.js [hunk 3]
```diff
@@ -14,27 +14,49 @@ describe('DeceasedDod', () => {
...
 
-    describe('handlePost()', () => {
-        let ctx;
-        let errors;
-        let formdata;
-        let session;
-        let hostname;
-        let featureToggles;
+    describe('getContextData()', () => {
+        it('should return the ctx with the deceased address and the screening_question feature toggle on', (done) => {
```
Comment (shepda) [inline test/unit/testDeceasedDod.js]: Please update the description for this test as it is for deceased dod but it is referring to deceased address.

File: test/unit/testDeceasedDod.js [hunk 4]
```diff
@@ -14,27 +14,49 @@ describe('DeceasedDod', () => {
...
+            });
+            done();
+        });
 
-        it('should return the ctx with the deceased dod and the screening_question feature toggle', (done) => {
-            ctx = {
-                dob_day: '02',
-                dob_month: '03',
-                dob_year: '1952'
+        it('should return the ctx with the deceased address and the screening_question feature toggle off', (done) => {
```
Comment (shepda) [inline test/unit/testDeceasedDod.js]: Please update the description for this test as it is for deceased dod but it is referring to deceased address.

File: test/unit/testDeceasedMarried.js [hunk 5]
```diff
@@ -14,23 +14,37 @@ describe('DeceasedMarried', () => {
...
 
-    describe('handlePost()', () => {
-        let ctx;
-        let errors;
-        let formdata;
-        let session;
-        let hostname;
-        let featureToggles;
+    describe('getContextData()', () => {
+        it('should return the ctx with the deceased address and the screening_question feature toggle on', (done) => {
```
Comment (shepda) [inline test/unit/testDeceasedMarried.js]: Please update the description for this test as it is for deceased married but it is referring to deceased address.

File: test/unit/testDeceasedMarried.js [hunk 6]
```diff
@@ -14,23 +14,37 @@ describe('DeceasedMarried', () => {
...
+                isToggleEnabled: true,
+                sessionID: 'dummy_sessionId'
+            });
+            done();
+        });
 
-        it('should return the ctx with the deceased married status and the screening_question feature toggle', (done) => {
-            ctx = {
-                married: 'Yes'
+        it('should return the ctx with the deceased address and the screening_question feature toggle off', (done) => {
```
Comment (shepda) [inline test/unit/testDeceasedMarried.js]: Please update the description for this test as it is for deceased married but it is referring to deceased address.

File: test/unit/testDeceasedName.js [hunk 7]
```diff
@@ -14,25 +14,41 @@ describe('DeceasedName', () => {
...
-        let errors;
-        let formdata;
-        let session;
-        let hostname;
-        let featureToggles;
-
-        it('should return the ctx with the deceased name and the screening_question feature toggle', (done) => {
-            ctx = {
+    describe('getContextData()', () => {
+        it('should return the ctx with the deceased address and the screening_question feature toggle on', (done) => {
```
Comment (shepda) [inline test/unit/testDeceasedName.js]: Please update the description for this test as it is for deceased name but it is referring to deceased address.

File: test/unit/testDeceasedName.js [hunk 8]
```diff
@@ -14,25 +14,41 @@ describe('DeceasedName', () => {
...
                 firstName: 'Deceased FN',
-                lastName: 'Deceased FN'
+                lastName: 'Deceased FN',
+                isToggleEnabled: true,
+                sessionID: 'dummy_sessionId'
+            });
+            done();
+        });
+
+        it('should return the ctx with the deceased address and the screening_question feature toggle off', (done) => {
```
Comment (shepda) [inline test/unit/testDeceasedName.js]: Please update the description for this test as it is for deceased name but it is referring to deceased address.

File: test/unit/testIhtPaper.js [hunk 9]
```diff
@@ -14,32 +14,45 @@ describe('IhtPaper', () => {
...
-        let errors;
-        let formdata;
-        let session;
-        let hostname;
-        let featureToggles;
-
-        it('should return the ctx with the deceased married status and the screening_question feature toggle', (done) => {
-            ctx = {
+    describe('getContextData()', () => {
+        it('should return the ctx with the deceased address and the screening_question feature toggle on', (done) => {
```
Comment (shepda) [inline test/unit/testIhtPaper.js]: Please update the description for this test as it is for iht paper but it is referring to deceased address.

File: test/unit/testIhtPaper.js [hunk 10]
```diff
@@ -14,32 +14,45 @@ describe('IhtPaper', () => {
...
                 grossIHT205: '500000',
-                netIHT205: '400000'
+                netIHT205: '400000',
+                isToggleEnabled: true,
+                sessionID: 'dummy_sessionId'
+            });
+            done();
+        });
+
+        it('should return the ctx with the deceased address and the screening_question feature toggle off', (done) => {
```
Comment (shepda) [inline test/unit/testIhtPaper.js]: Please update the description for this test as it is for iht paper but it is referring to deceased address.

File: test/unit/testIhtValue.js [hunk 11]
```diff
@@ -14,27 +14,41 @@ describe('IhtValue', () => {
...
-        let errors;
-        let formdata;
-        let session;
-        let hostname;
-        let featureToggles;
-
-        it('should return the ctx with the deceased married status and the screening_question feature toggle', (done) => {
-            ctx = {
+    describe('getContextData()', () => {
+        it('should return the ctx with the deceased address and the screening_question feature toggle on', (done) => {
```
Comment (shepda) [inline test/unit/testIhtValue.js]: Please update the description for this test as it is for iht value but it is referring to deceased address.

File: test/unit/testIhtValue.js [hunk 12]
```diff
@@ -14,27 +14,41 @@ describe('IhtValue', () => {
...
                 grossValueOnline: '500000',
-                netValueOnline: '400000'
+                netValueOnline: '400000',
+                isToggleEnabled: true,
+                sessionID: 'dummy_sessionId'
+            });
+            done();
+        });
+
+        it('should return the ctx with the deceased address and the screening_question feature toggle off', (done) => {
```
Comment (shepda) [inline test/unit/testIhtValue.js]: Please update the description for this test as it is for iht value but it is referring to deceased address.

File: test/unit/testWillCodicils.js [hunk 13]
```diff
@@ -15,23 +15,37 @@ describe('WillCodicils', () => {
...
 
-    describe('handlePost()', () => {
-        let ctx;
-        let errors;
-        let formdata;
-        let session;
-        let hostname;
-        let featureToggles;
+    describe('getContextData()', () => {
+        it('should return the ctx with the deceased address and the screening_question feature toggle on', (done) => {
```
Comment (shepda) [inline test/unit/testWillCodicils.js]: Please update the description for this test as it is for will codicils but it is referring to deceased address.

File: test/unit/testWillCodicils.js [hunk 14]
```diff
@@ -15,23 +15,37 @@ describe('WillCodicils', () => {
...
+                isToggleEnabled: true,
+                sessionID: 'dummy_sessionId'
+            });
+            done();
+        });
 
-        it('should return the ctx with the deceased married status and the screening_question feature toggle', (done) => {
-            ctx = {
-                codicils: 'Yes'
+        it('should return the ctx with the deceased address and the screening_question feature toggle off', (done) => {
```
Comment (shepda) [inline test/unit/testWillCodicils.js]: Please update the description for this test as it is for will codicils but it is referring to deceased address.

File: test/unit/testWillCodicilsNumber.js [hunk 15]
```diff
@@ -14,23 +14,37 @@ describe('CodicilsNumber', () => {
...
 
-    describe('handlePost()', () => {
-        let ctx;
-        let errors;
-        let formdata;
-        let session;
-        let hostname;
-        let featureToggles;
+    describe('getContextData()', () => {
+        it('should return the ctx with the deceased address and the screening_question feature toggle on', (done) => {
```
Comment (shepda) [inline test/unit/testWillCodicilsNumber.js]: Please update the description for this test as it is for will codicils number but it is referring to deceased address.

File: test/unit/testWillCodicilsNumber.js [hunk 16]
```diff
@@ -14,23 +14,37 @@ describe('CodicilsNumber', () => {
...
+                isToggleEnabled: true,
+                sessionID: 'dummy_sessionId'
+            });
+            done();
+        });
 
-        it('should return the ctx with the deceased married status and the screening_question feature toggle', (done) => {
-            ctx = {
-                codicilsNumber: '3'
+        it('should return the ctx with the deceased address and the screening_question feature toggle off', (done) => {
```
Comment (shepda) [inline test/unit/testWillCodicilsNumber.js]: Please update the description for this test as it is for will codicils number but it is referring to deceased address.

File: app/steps/ui/will/codicilsnumber/index.js [hunk 17]
```diff
@@ -12,16 +12,10 @@ class CodicilsNumber extends ValidationStep {
     getContextData(req) {
         const ctx = super.getContextData(req);
         ctx.codicilsNumber = ctx.codicilsNumber ? parseInt(ctx.codicilsNumber): ctx.codicilsNumber;
+        ctx.isToggleEnabled = FeatureToggle.isEnabled(req.session.featureToggles, 'screening_questions');
         return ctx;
     }
 
-    handlePost(ctx, errors, formdata, session, hostname, featureToggles) {
-        ctx.codicilsNumber = ctx.codicilsNumber || 0;
```
[the remark this answers is not in the document]
Comment (LucaDelBuonoHMCTS) [reply app/steps/ui/will/codicilsnumber/index.js]: My mistake, I should only have removed the line about the feature toggle, now `handlePost` is back

File: app/steps/ui/will/codicilsnumber/index.js [hunk 18]
```diff
@@ -16,6 +16,11 @@ class CodicilsNumber extends ValidationStep {
         return ctx;
     }
 
+    handlePost(ctx, errors) {
+        ctx.codicilsNumber = ctx.codicilsNumber || 0;
+        return [ctx, errors];
+    }
```
Comment (shepda) [inline app/steps/ui/will/codicilsnumber/index.js]: Can we please have unit tests for this method, we'll need one for when `ctx.codicilsNumber` exists and one for when `ctx.codicilsNumber` doesn't exist.

Comment (LucaDelBuonoHMCTS) [reply app/steps/ui/will/codicilsnumber/index.js]: Argh, doing 3 things at the same time on a Friday afternoon doesn't help :)
Added.

File: test/unit/testWillCodicilsNumber.js [hunk 19]
```diff
@@ -14,23 +14,64 @@ describe('CodicilsNumber', () => {
...
+            const ctx = CodicilsNumber.getContextData(req);
+            expect(ctx).to.deep.equal({
+                codicilsNumber: 3,
+                isToggleEnabled: true,
+                sessionID: 'dummy_sessionId'
+            });
+            done();
+        });
+
+        it('should return the ctx with the will codicils number and the screening_question feature toggle off', (done) => {
```
[the remark this answers is not in the document]
Comment (LucaDelBuonoHMCTS) [reply test/unit/testWillCodicilsNumber.js]: Done

**Pull request closed without merging**